tests/integration_tests/performance/test_network_latency.py

```diff
@@ -0,0 +1,199 @@
+                                                                   delta)
+    return criteria
+
+
+def measurements():
+    """Define the produced measurements."""
+    ms_defs = [consumer.MeasurementDef(LATENCY, "millisecond"),
+               consumer.MeasurementDef(PKT_LOSS, "percentage")]
+    return ms_defs
+
+
+def stats():
+    """Define statistics based on the measurements."""
+    criteria = pass_criteria()
+    # Add default statistics for "latency" measurement.
+    stats_defs = consumer.StatisticDef.defaults(LATENCY,
+                                                pass_criteria=criteria)
+    stats_defs.append(consumer.StatisticDef(PKT_LOSS, PKT_LOSS,
+                                            consumer.Method.last))
+    return stats_defs
+
+
+def consume_ping_output(cons, raw_data, requests):
+    """Consume ping output.
+
+    Output example:
+    PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
+    64 bytes from 8.8.8.8: icmp_seq=1 ttl=118 time=17.7 ms
+    64 bytes from 8.8.8.8: icmp_seq=2 ttl=118 time=17.7 ms
+    64 bytes from 8.8.8.8: icmp_seq=3 ttl=118 time=17.4 ms
+    64 bytes from 8.8.8.8: icmp_seq=4 ttl=118 time=17.8 ms
+
+    --- 8.8.8.8 ping statistics ---
+    4 packets transmitted, 4 received, 0% packet loss, time 3005ms
+    rtt min/avg/max/mdev = 17.478/17.705/17.808/0.210 ms
+    """
+    eager_map(cons.consume_measurement_def, measurements())
+    eager_map(cons.consume_stat_def, stats())
+
+    st_keys = [consumer.StatisticDef.MIN_KEY,
+               consumer.StatisticDef.AVG_KEY,
+               consumer.StatisticDef.MAX_KEY,
+               consumer.StatisticDef.STDDEV_KEY]
+
+    output = raw_data.strip().split('\n')
+    assert len(output) > 2
+
+    # E.g: rtt min/avg/max/mdev = 17.478/17.705/17.808/0.210 ms
+    stat_values = output[-1]
+    stat_values = re.findall(FLOAT_REG, stat_values)
+    assert len(stat_values) == 4
+    for index, stat_value in enumerate(stat_values[:4]):
+        cons.consume_result(name=st_keys[index], value=float(stat_value))
+
+    # E.g: 4 packets transmitted, 4 received, 0% packet loss, time 3005ms
+    packet_stats = output[-2]
+    packet_stats = packet_stats.split(',')[2]
+    packet_stats = re.findall(INT_REG, packet_stats)
+    assert len(packet_stats) == 1
+    cons.consume_result(name=PKT_LOSS, value=packet_stats[0])
+
+    # Compute percentiles.
+    seqs = output[1:requests + 1]
+    times = list()
+    for index, seq in enumerate(seqs):
+        time = re.findall(FLOAT_REG + ' ms', seq)[0]
+        time = re.findall(FLOAT_REG, time)[0]
+        times.append(time)
+
+    times.sort()
+    cons.consume_result(name=consumer.StatisticDef.P50_KEY,
+                        value=times[int(requests * 0.5)])
+    cons.consume_result(name=consumer.StatisticDef.P90_KEY,
+                        value=times[int(requests * 0.9)])
+    cons.consume_result(name=consumer.StatisticDef.P99_KEY,
+                        value=times[int(requests * 0.99)])
```

Comment:
Shouldn't the consumer or aggregator compute this automatically from the measurements that we supply ?

Reply:
Generally, there are two modes of driving the producer/consumer pipes. Some of them can produce *measurements data points*, which can be transformed in statistics by applying  on them statistical methods in the consumer statistics generation step while others can directly *produce the statistics data*, which should be directly forwarded to the consumer statistics generation step. However, both measurements and statistics definitions are needed in either way for statistics generation, since statistics units are set by measurements definitions.

In this specific case we initialize the `LambdaConsumer` with `consume_stats=True`. This means we directly consume statistics for our measurements, rather than measurements data.

On a separate note, I must avoid asserting on the measurements consumption expectancy, from `Consumer::process` method: i.e if the consumer is initialized with `consume_stats=True` than we must not check if measurements definitions have corresponding data points consumed. I'll let you know after I fix it.